File: backend/src/services/authService.js
/**
 * Service d'authentification
 * Gère l'inscription, la connexion et la gestion des utilisateurs
 */

const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const db = require('../config/database');
const config = require('../config/config');
const { AppError } = require('../middleware/errorHandler');

/**
 * Inscription d'un nouvel utilisateur
 */
const register = async (userData) => {
  const { email, password, full_name, phone, role = 'user' } = userData;

  // Vérifier si l'email existe déjà
  const existingUser = await db.query('SELECT id FROM users WHERE email = $1', [
    email,
  ]);

  if (existingUser.rows.length > 0) {
    throw new AppError('Cet email est déjà utilisé', 400);
  }

  // Hasher le mot de passe
  const saltRounds = 10;
  const passwordHash = await bcrypt.hash(password, saltRounds);

  // Insérer le nouvel utilisateur
  const result = await db.query(
    `INSERT INTO users (email, password_hash, full_name, phone, role) 
     VALUES ($1, $2, $3, $4, $5) 
     RETURNING id, email, full_name, phone, role, created_at`,
    [email, passwordHash, full_name, phone, role]
  );

  const user = result.rows[0];

  // Générer le token JWT
  const token = generateToken(user);

  return {
    user: {
      id: user.id,
      email: user.email,
      full_name: user.full_name,
      phone: user.phone,
      role: user.role,
      created_at: user.created_at,
    },
    token,
  };
};

/**
 * Connexion d'un utilisateur
 */
const login = async (email, password) => {
  // Récupérer l'utilisateur
  const result = await db.query(
    'SELECT id, email, password_hash, full_name, phone, role, created_at FROM users WHERE email = $1',
    [email]
  );

  if (result.rows.length === 0) {
    throw new AppError('Email ou mot de passe incorrect', 401);
  }

  const user = result.rows[0];

  // Vérifier le mot de passe
  const isPasswordValid = await bcrypt.compare(password, user.password_hash);

  if (!isPasswordValid) {
    throw new AppError('Email ou mot de passe incorrect', 401);
  }

  // Générer le token JWT
  const token = generateToken(user);

  return {
    user: {
      id: user.id,
      email: user.email,
      full_name: user.full_name,
      phone: user.phone,
      role: user.role,
      created_at: user.created_at,
    },
    token,
  };
};

/**
 * Récupérer le profil d'un utilisateur
 */
const getProfile = async (userId) => {
  const result = await db.query(
    'SELECT id, email, full_name, phone, role, created_at FROM users WHERE id = $1',
    [userId]
  );

  if (result.rows.length === 0) {
    throw new AppError('Utilisateur non trouvé', 404);
  }

  return result.rows[0];
};

/**
 * Mettre à jour le profil d'un utilisateur
 */
const updateProfile = async (userId, updateData) => {
  const { full_name, phone } = updateData;

  const result = await db.query(
    `UPDATE users 
     SET full_name = COALESCE($1, full_name), 
         phone = COALESCE($2, phone),
         updated_at = CURRENT_TIMESTAMP
     WHERE id = $3 
     RETURNING id, email, full_name, phone, role, created_at, updated_at`,
    [full_name, phone, userId]
  );

  if (result.rows.length === 0) {
    throw new AppError('Utilisateur non trouvé', 404);
  }

  return result.rows[0];
};

/**
 * Changer le mot de passe
 */
const changePassword = async (userId, currentPassword, newPassword) => {
  // Récupérer l'utilisateur
  const result = await db.query(
    'SELECT password_hash FROM users WHERE id = $1',
    [userId]
  );

  if (result.rows.length === 0) {
    throw new AppError('Utilisateur non trouvé', 404);
  }

  const user = result.rows[0];

  // Vérifier le mot de passe actuel
  const isPasswordValid = await bcrypt.compare(
    currentPassword,
    user.password_hash
  );

  if (!isPasswordValid) {
    throw new AppError('Mot de passe actuel incorrect', 401);
  }

  // Hasher le nouveau mot de passe
  const saltRounds = 10;
  const newPasswordHash = await bcrypt.hash(newPassword, saltRounds);

  // Mettre à jour le mot de passe
  await db.query(
    'UPDATE users SET password_hash = $1, updated_at = CURRENT_TIMESTAMP WHERE id = $2',
    [newPasswordHash, userId]
  );

  return { message: 'Mot de passe modifié avec succès' };
};

/**
 * Générer un token JWT
 */
const generateToken = (user) => {
  return jwt.sign(
    {
      id: user.id,
      email: user.email,
      role: user.role,
    },
    config.jwt.secret,
    { expiresIn: config.jwt.expiresIn }
  );
};

module.exports = {
  register,
  login,
  getProfile,
  updateProfile,
  changePassword,
};

File: frontend/lib/data/services/auth_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import '../../core/constants/api_constants.dart';
import '../../core/constants/storage_constants.dart';
import '../../domain/models/user.dart';
import 'api_service.dart';

/// Service d'authentification
/// Gère login, register, profile, logout avec JWT
class AuthService {
  final ApiService _apiService;
  final SharedPreferences _prefs;

  AuthService(this._apiService, this._prefs);

  /// Inscription d'un nouvel utilisateur
  Future<AuthResponse> register({
    required String name,
    required String email,
    required String password,
  }) async {
    final response = await _apiService.post(
      ApiConstants.register,
      data: {
        'full_name': name, // Backend attend 'full_name' et non 'name'
        'email': email,
        'password': password,
      },
    );

    // L'API retourne les données dans response.data.data
    final authResponse = AuthResponse.fromJson(response.data['data'] as Map<String, dynamic>);
    await _saveAuthData(authResponse);
    return authResponse;
  }

  /// Connexion utilisateur
  Future<AuthResponse> login({
    required String email,
    required String password,
  }) async {
    final response = await _apiService.post(
      ApiConstants.login,
      data: {
        'email': email,
        'password': password,
      },
    );

    // L'API retourne les données dans response.data.data
    final authResponse = AuthResponse.fromJson(response.data['data'] as Map<String, dynamic>);
    await _saveAuthData(authResponse);
    return authResponse;
  }

  /// Récupérer le profil de l'utilisateur connecté
  Future<User> getProfile() async {
    final response = await _apiService.get(ApiConstants.profile);
    // L'API retourne le user directement dans response.data
    return User.fromJson(response.data as Map<String, dynamic>);
  }

  /// Mettre à jour le profil
  Future<User> updateProfile({
    String? name,
    String? email,
    String? phone,
    String? address,
  }) async {
    final data = <String, dynamic>{};
    if (name != null) data['name'] = name;
    if (email != null) data['email'] = email;
    if (phone != null) data['phone'] = phone;
    if (address != null) data['address'] = address;

    final response = await _apiService.put(
      ApiConstants.updateProfile,
      data: data,
    );

    // L'API retourne le user dans response.data
    final user = User.fromJson(response.data as Map<String, dynamic>);
    await _saveUser(user);
    return user;
  }

  /// Changer le mot de passe
  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    await _apiService.put(
      ApiConstants.changePassword,
      data: {
        'current_password': currentPassword,
        'new_password': newPassword,
      },
    );
  }

  /// Déconnexion
  Future<void> logout() async {
    await _prefs.remove(StorageConstants.authToken);
    await _prefs.remove(StorageConstants.userRole);
    await _prefs.remove(StorageConstants.userFullName);
    await _prefs.remove(StorageConstants.userEmail);
    await _prefs.remove(StorageConstants.userRole);
  }

  /// Vérifier si l'utilisateur est connecté
  bool isLoggedIn() {
    return _prefs.getString(StorageConstants.authToken) != null;
  }

  /// Vérifier si l'utilisateur est admin
  bool isAdmin() {
    return _prefs.getString(StorageConstants.userRole) == 'admin';
  }

  /// Récupérer l'utilisateur depuis le cache local
  User? getCachedUser() {
    final userId = _prefs.getInt(StorageConstants.userId);
    final fullName = _prefs.getString(StorageConstants.userFullName);
    final email = _prefs.getString(StorageConstants.userEmail);
    final role = _prefs.getString(StorageConstants.userRole);

    if (userId == null || email == null || role == null) {
      return null;
    }

    return User(
      id: userId,
      fullName: fullName ?? '',
      email: email,
      role: role,
      phone: _prefs.getString(StorageConstants.userPhone),
    );
  }

  /// Sauvegarder les données d'authentification
  Future<void> _saveAuthData(AuthResponse authResponse) async {
    await _prefs.setString(StorageConstants.authToken, authResponse.token);
    await _saveUser(authResponse.user);
  }

  /// Sauvegarder l'utilisateur en cache
  Future<void> _saveUser(User user) async {
    await _prefs.setInt(StorageConstants.userId, user.id);
    await _prefs.setString(StorageConstants.userFullName, user.fullName);
    await _prefs.setString(StorageConstants.userEmail, user.email);
    await _prefs.setString(StorageConstants.userRole, user.role);
    if (user.phone != null) {
      await _prefs.setString(StorageConstants.userPhone, user.phone!);
    }
  }
}
